### ISSUE TITLE
Allow user-defined attributes using `dirvish-define-attribute`

### ISSUE THREAD

**Issue by Zander671** [alexluigit/dirvish#286]
### Thank you for the bug report

- [X] I am using the latest version of `dirvish` related packages.
- [X] I checked [CUSTOMIZING|EXTENSIONS](https://github.com/alexluigit/dirvish#resources).
- [ ] You may also try reproduce the issue using clean environment and minimal configurations with
the command `emacs -Q`.

### Bug description

The function `dirvish--check-deps` removes entries from `dirvish-attributes` unless they have a corresponding entry in `dirvish-libraries`. For example, I defined the following attribute in my configuration:
```elisp
(dirvish-define-attribute file-owner-mode
  "The file's owner and mode."
  :index 2
  :when (and (dirvish-prop :root) dired-hide-details-mode
             (> win-width 60))
  (let ((root (dirvish-prop :root))
        (uid (file-attribute-user-id f-attrs)))
    (unless (or (dirvish-prop :remote) (stringp uid))
      (unless (and (equal root (car my/-dirvish-uid-name-cache))
                   (hash-table-p (cdr my/-dirvish-uid-name-cache)))
        (setq my/-dirvish-uid-name-cache
              (cons root (make-hash-table :test 'equal))))
      (if-let ((name (gethash uid (cdr my/-dirvish-uid-name-cache))))
          (setq uid name)
        (let* ((new-attrs (file-attributes f-name 'string))
               (new-name (file-attribute-user-id new-attrs)))
          (puthash uid new-name
                   (cdr my/-dirvish-uid-name-cache))
          (setq uid new-name))))
    (cons 'right (propertize (format "%s %s" uid (file-attribute-modes f-attrs))
                             'face (or hl-face 'dirvish-file-time)))))
```
however, this attribute would not show up until I used the transient menu and toggled one of the other attributes. I was, however, able to fix this problem by adding the following line to my configuration as a workaround:
```elisp
(add-to-list 'dirvish-libraries '(dirvish file-owner-mode))
```
This is because `dirvish--check-deps` (which sets `dirvish--working-attrs`) only actually process attributes that have a corresponding entry in `dirvish-librarires`. This is due to the following lines:
```elisp
   for (lib . feat) in dirvish-libraries do
   (let ((m-attr (cl-intersection feat dirvish-attributes))
         (feat-in-lib (cl-intersection feat feat-reqs)))
     (when (or m-attr feat-in-lib) (require lib))
     (and m-attr (setq attrs (append attrs m-attr))))
``` 
and
```elsip
(setf dirvish--working-attrs (dirvish--attrs-expand attrs))
```

### Steps to reproduce

It seems to me than any module defined with `dirvish-define-attribute` that does not have a corresponding entry in `dirvish-libraries` will have this problem.

### Expected behavior

Probably `dirvish--check-deps` should be rewritten to not require modules to have an entry in `dirvish-libraries` (I can try to implement this and make a pull request if you want). Alternatively, this should at least be documented in `dirvish-define-attribute` so that users know they also need to add an entry to `dirvish-libraries`.

### OS

Linux

### Emacs Version

29

### Emacs Configurations

https://git.zander.im/zander671/emacs-config

**Comment by alexluigit** [comment]
Thanks for the report and sorry for the late response! I do think `dirvish-libraries` is not ideal, there definitely a smarter of doing this, maybe quering the `obarray`?  I'd love to hear your feedback, also I'm happy to review your PR if you still willing to do so.

**Comment by Zander671** [comment]
One option that is minimally intrusive is something like this:
```elisp
(defun dirvish--attrs-expand (attrs)
  "Expand ATTRS to `dirvish--working-attrs'."
  (sort (cl-loop for attr in attrs
                 for lst = (alist-get attr dirvish--available-attrs)
                 for (idx width pred render ov _) = lst
                 when lst
                 collect (list idx (eval width) pred render ov)
                 else do (warn "[Dirvish]: unknown attribute: %s" attr))
        (lambda (a b) (< (car a) (car b)))))

(defun dirvish--check-deps ()
  "Remove invalid widgets, raise warnings for missing dependencies."
  (cl-loop
   with (m . h) = (cons dirvish-mode-line-format dirvish-header-line-format)
   with (ml-l . ml-r) = (cons (plist-get m :left) (plist-get m :right))
   with (hl-l . hl-r) = (cons (plist-get h :left) (plist-get h :right))
   with feat-reqs = (append dirvish-preview-dispatchers ml-l ml-r hl-l hl-r)
   with attrs = (apply 'list 'hl-line 'symlink-target dirvish-attributes)
   for (lib . feat) in dirvish-libraries do
   (let ((m-attr (cl-intersection feat dirvish-attributes))
         (feat-in-lib (cl-intersection feat feat-reqs)))
     (when (or m-attr feat-in-lib) (require lib)))
   finally
   (setf dirvish--mode-line-fmt (dirvish--mode-line-composer ml-l ml-r))
   (setf dirvish--header-line-fmt (dirvish--mode-line-composer hl-l hl-r t))
   (setf dirvish--working-preview-dispathchers (dirvish--preview-dps-validate))
   (setf dirvish--working-attrs (dirvish--attrs-expand attrs))))
```

With the above change, the `dirvish-attributes` list not touched. This means means that any errors in the list will alert the user to something being wrong, rather than just being silently removed. In addition, with the above code built-in attributes that need a file to be `require`d are still processed as usual. If the user puts an unknown entry in `dirvish-attributes` it will now raise a warning in `dirvish--attrs-expand`.

Weather or not to remove `dirvish-libraries` entirely is a different question. One possibility would indeed be to use `obarray` and `put` some kind of property on the symbol for each feature. However, this might be harder to debug than the status quo, and there is always the chance of one of the feature symbols being `uninterned` for some reason (though probably unlikely).

**Comment by alexluigit** [comment]
> This means means that any errors in the list will alert the user to something being wrong, rather than just being silently removed. In addition, with the above code built-in attributes that need a file to be `require`d are still processed as usual. If the user puts an unknown entry in `dirvish-attributes` it will now raise a warning in `dirvish--attrs-expand`.

Yeah it is more user friendly, but we really need is to include defined attributes here, regardless they are in-built or not.

> Weather or not to remove `dirvish-libraries` entirely is a different question. One possibility would indeed be to use `obarray` and `put` some kind of property on the symbol for each feature. However, this might be harder to debug than the status quo, and there is always the chance of one of the feature symbols being `uninterned` for some reason (though probably unlikely).

Maybe we can just autoload all attribute functions/preview-dispatchers?  This is already being done with all the transient menu functions.

**Comment by Zander671** [comment]
> Yeah it is more user friendly, but we really need is to include defined attributes here, regardless they are in-built or not.

For user modules, I don't know how much of a problem this is as most users  are likely to define features in a `with-eval-after-load` (or with `use-package` to the same effect) at the same time they add the entry to `dirvish-attributes`, etc. For the users that want to lazy load their custom features and for other packages that want to provide their own Drivish features, Dirvish could provide a function to the effect of
```elisp
(defun dirvish-defer-features (lib &rest features)
    "Define each of FEATUES to be loaded from LIB.
Each symbol in FEATUES is taken to be a Drivish feature (defined with
`dirvish-define-attribute', `dirvish-define-mode-line', or
`dirvish-define-preview').  Dirvish will arrange for LIB to be `require'd when
the time comes for each of the FEATURES to be used."
    (dolist (feature features)
      (cl-pushnew feature (alist-get lib dirvish-libraries))))
```
packages building on Dirvish can then use
```elisp
;;;###autoload (dirvish-defer-features 'dirvish-foo 'dirvish-foo-attribute 'dirvish-foo-modeline)
```
to add their entries to `dirvish-libraries`.

> Maybe we can just autoload all attribute functions/preview-dispatchers? This is already being done with all the transient menu functions. So we automatically take care of all in-built widgets, for user defined widgets all we have to do is to check if the expanded function name exists.

Even now, I think that if you put an autoload comment above any of these definitions, it will be picked up by `loaddefs-generate` and the entire content of the macro will be called when Emacs starts. However, this requires that the file have loaddefs generated for it, which is not likely to happen for user files or packages not installed with `package.el`, `straight.el`, etc.

**Comment by alexluigit** [comment]
> For the users that want to lazy load their custom features and for other packages that want to provide their own Drivish features, Dirvish could provide a function to the effect of
> 
> (defun dirvish-defer-features (lib &rest features)
>     "Define each of FEATUES to be loaded from LIB.
> Each symbol in FEATUES is taken to be a Drivish feature (defined with
> `dirvish-define-attribute', `dirvish-define-mode-line', or
> `dirvish-define-preview').  Dirvish will arrange for LIB to be `require'd when
> the time comes for each of the FEATURES to be used."
>     (dolist (feature features)
>       (cl-pushnew feature (alist-get lib dirvish-libraries))))
> packages building on Dirvish can then use
> 
> ;;;###autoload (dirvish-defer-features 'dirvish-foo 'dirvish-foo-attribute 'dirvish-foo-modeline)
> to add their entries to `dirvish-libraries`.
> 

This looks neat! Would you mind open a PR? We may also need to update documentations in [CUSTOMIZING.org](https://github.com/alexluigit/dirvish/blob/main/docs/CUSTOMIZING.org), perhaps insert a new section between **File Preview** and **Sample Config** (just a suggestion, we can discuss it afterwards).
